### PR TITLE
[Anima] Add img2img pipeline blocks

### LINKED ISSUE #13903: [Anima] Add img2img capability
**Is your feature request related to a problem? Please describe.**
Anima in diffusers is a modular diffusers only model but is still missing the image to image functionality. This should be relatively easy to add and can help people understand how to work with modular diffusers. 

This feature request is to see if someone from the community is willing to contribute this functionality. Also it would help use if they see something we can improve with modular diffusers.

### PR DESCRIPTION
**Pull request by PreethamNoelP**
# What does this PR do?

This PR adds image-to-image (`img2img`) support to the Anima modular pipeline, as requested in #13903. The implementation follows the z_image modular pattern: img2img is a workflow inside the unified `AnimaAutoBlocks`, selected automatically when an `image` (or pre-encoded `image_latents`) is provided.

## Design

**New blocks**

- `AnimaImg2ImgVaeEncoderStep` — preprocesses the input image and encodes it with the VAE, returning `image_latents` only. Encoding is independent of `strength` / `num_inference_steps` / `num_images_per_prompt`, so latents encoded once can be reused across settings.
- `AnimaImageInputStep` — input processing step (analogous to `AnimaTextInputStep`) that expands `image_latents` to the final denoising batch (`batch_size * num_images_per_prompt`) and derives `height`/`width` from the latents when not provided.
- `AnimaImg2ImgSetTimestepsStep` — computes the full timestep schedule, then slices it by `strength` via `get_timesteps()`, which also sets the scheduler's begin index.
- `AnimaImg2ImgPrepareLatentsStep` — generates noise and mixes it with the image latents via `scheduler.scale_noise()` at the first sliced timestep, and creates the Cosmos padding mask.
- `AnimaImg2ImgCoreDenoiseStep` — flat `SequentialPipelineBlocks`: `text_conditioning → input → image_input → set_timesteps → prepare_latents → denoise`, mirroring `ZImageImage2ImageCoreDenoiseStep`.
- `AnimaAutoVaeImageEncoderStep` — `AutoPipelineBlocks` triggered by `image`; skipped for text2image.
- `AnimaAutoDenoiseStep` — `AutoPipelineBlocks` with `block_trigger_inputs = ["image_latents", None]`, selecting img2img or text2image core denoise.

**Unified pipeline**

`AnimaAutoBlocks` = `text_encoder → vae_encoder (auto) → denoise (auto) → decode`, with a `_workflow_map` covering both `text2image` and `img2img`. No separate img2img pipeline class.

## Usage

```python
import torch
from diffusers import AnimaModularPipeline
from diffusers.modular_pipelines.anima import AnimaAutoBlocks

pipe = AnimaModularPipeline(
    blocks=AnimaAutoBlocks(),
    pretrained_model_name_or_path="circlestone-labs/Anima-Base-v1.0-Diffusers",
)
pipe.load_components(torch_dtype=torch.bfloat16)
pipe.to("cuda")

image = pipe(
    prompt="masterpiece, best quality, mountain landscape at golden hour, snow-capped peaks, pine forest, dramatic clouds, cinematic, detailed",
    image=input_image,
    strength=0.75,
    num_inference_steps=28,
    generator=torch.Generator(device="cuda").manual_seed(42),
).images[0]
```

## Test results

```
pytest tests/modular_pipelines/anima/ -q
28 passed, 10 skipped
```

Fixes #13903

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Link: #13903
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@asomoza @yiyixuxu

### CODE REVIEW

**Comment by asomoza** [comment]
thanks @PreethamNoelP, can we start with updating the PR with a code snippet to test it and the images to see that it works before doing the review.

**Comment by PreethamNoelP** [comment]
Hi @asomoza, here is the code snippet and output images.

Usage:
```
import torch
from diffusers import AnimaModularPipeline
from diffusers.modular_pipelines.anima import AnimaImg2ImgAutoBlocks

pipe = AnimaModularPipeline(
    blocks=AnimaImg2ImgAutoBlocks(),
    pretrained_model_name_or_path="circlestone-labs/Anima-Base-v1.0-Diffusers",
)
pipe.load_components(torch_dtype=torch.bfloat16)
pipe.to("cuda")

image = pipe(
    prompt="masterpiece, best quality, mountain landscape at golden hour, snow-capped peaks, pine forest, dramatic clouds, cinematic, detailed",
    image=input_image,
    strength=0.75,
    num_inference_steps=28,
    generator=torch.Generator(device="cuda").manual_seed(42),
).images[0]

```

Outputs:
side_by_side_comparison

File: src/diffusers/modular_pipelines/anima/before_denoise.py [hunk 1]
```diff
@@ -414,3 +427,77 @@ def __call__(self, components: AnimaModularPipeline, state: PipelineState) -> Pi
 
         self.set_block_state(state, block_state)
         return components, state
+
+
+# Copied from diffusers.modular_pipelines.anima.before_denoise.AnimaSetTimestepsStep
+class AnimaImg2ImgSetTimestepsStep(ModularPipelineBlocks):
```
Comment (asomoza) [inline src/diffusers/modular_pipelines/anima/before_denoise.py]: you used the copied from here but you changed the function no? You can catch these errors and the code quality by following the [contributing](https://github.com/huggingface/diffusers/blob/main/docs/source/en/conceptual/contribution.md) doc

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py [hunk 2]
```diff
@@ -181,3 +183,144 @@ def description(self) -> str:
...
+            "VAE encoding runs after AnimaTextInputStep so batch_size is available in state."
+        )
+
+    @property
+    def outputs(self):
+        return [OutputParam.template("latents")]
+
+
+# auto_docstring
+class AnimaImg2ImgAutoBlocks(SequentialPipelineBlocks):
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/modular_blocks_anima.py]: we do not need a `AnimaImg2ImgAutoBlocks`. We add a workflow into the `AnimaAutoBlocks`

see example https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/z_image/modular_blocks_z_image.py#L325

File: src/diffusers/modular_pipelines/anima/before_denoise.py (hunk 1 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/before_denoise.py]: Removed the # Copied from tag above AnimaImg2ImgSetTimestepsStep. 
Since the class intentionally skips the set_begin_index(0) call (that's handled later by the VAE encoder step), it was never truly identical to the source, so the tag shouldn't have been there.
Please let me know if any further changes are required.

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py (hunk 2 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/modular_blocks_anima.py]: Removed AnimaImg2ImgAutoBlocks entirely and folded img2img into AnimaAutoBlocks following the z_image pattern. 
I added AnimaAutoDenoiseStep (an AutoPipelineBlocks that picks between img2img and txt2img based on whether image is provided) and AnimaImg2ImgDenoiseStep (a SequentialPipelineBlocks that wraps the img2img-specific set_timesteps + denoise steps). 
AnimaAutoBlocks now has a _workflow_map covering both workflows, so users only need one pipeline - passing image= automatically triggers img2img.
Please let me know if any further changes are required.

**Comment by asomoza** [comment]
@PreethamNoelP thanks, sorry I missed the update, I checked it and it looks good to me, I tested it and the outputs seems coherent too.

@yiyixuxu can you review it one last time to see if I missed something?

**Review by asomoza — APPROVED** [review]

**Comment by asomoza** [comment]
@bot /style

**Comment by ghunkins** [comment]
Any updates on this by chance? Would love to see this in `main`! cc @yiyixuxu

File: src/diffusers/modular_pipelines/anima/encoders.py [hunk 3]
```diff
@@ -251,3 +256,235 @@ def __call__(self, components: AnimaModularPipeline, state: PipelineState) -> Pi
...
+
+    @property
+    def description(self) -> str:
+        return (
+            "VAE Encoder step for Anima image-to-image generation. Encodes the input image, "
+            "slices the timestep schedule by strength, and adds noise via scheduler.scale_noise()."
+        )
+
+    @property
+    def inputs(self) -> list[InputParam]:
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/encoders.py:365]: the encoder step should just take an `image` (height/width too) and return image_latents, 

i.e. if you have already encoded the image once, you should not need to encode again just to use a different `strength`/`num_images_per_prompt` or `num_inference_steps` etc

see https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/z_image/encoders.py#L272

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py [hunk 4]
```diff
@@ -122,18 +123,228 @@ def outputs(self):
...
+              The conditional model inputs for the Anima denoiser.
+
+      Outputs:
+          latents (`Tensor`):
+              Denoised latents.
+    """
+
+    block_classes = [AnimaImg2ImgDenoiseStep, AnimaCoreDenoiseStep]
+    block_names = ["img2img", "text2image"]
+    block_trigger_inputs = ["image", None]
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/modular_blocks_anima.py]:
```suggestion
    block_trigger_inputs = ["image_latents", None]
```

File: src/diffusers/modular_pipelines/anima/encoders.py (hunk 3 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/encoders.py:365]: Hi @yiyixuxu, thank you for the suggestion. AnimaImg2ImgVaeEncoderStep has been refactored to only handle image encoding, returning image_latents. The timestep slicing and noise mixing logic has been extracted into a new AnimaImg2ImgPrepareLatentsStep, following the same separation as ZImageVaeImageEncoderStep and ZImagePrepareLatentswithImageStep.

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py (hunk 4 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/modular_blocks_anima.py]: Hi @yiyixuxu, thank you for the suggestion. AnimaAutoDenoiseStep has been updated to use block_trigger_inputs = ["image_latents", None]. Additionally, AnimaAutoVaeImageEncoderStep (triggered by "image") has been added as a sequential step in AnimaAutoBlocks, ensuring image_latents is placed in state before the denoise step selects the appropriate block - consistent with the ZImageAutoBlocks structure.

File: src/diffusers/modular_pipelines/anima/before_denoise.py [hunk 5]
```diff
@@ -414,3 +427,201 @@ def __call__(self, components: AnimaModularPipeline, state: PipelineState) -> Pi
...
+    @torch.no_grad()
+    def __call__(self, components: AnimaModularPipeline, state: PipelineState) -> PipelineState:
+        block_state = self.get_block_state(state)
+
+        device = components._execution_device
+        dtype = block_state.dtype if block_state.dtype is not None else components.vae.dtype
+
+        block_state.timesteps, block_state.num_inference_steps = get_timesteps(
+            components.scheduler, block_state.num_inference_steps, block_state.strength
+        )
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/before_denoise.py]: can we move this into `AnimaImg2ImgSetTimestepsStep`? makes more sense there

File: src/diffusers/modular_pipelines/anima/before_denoise.py [hunk 6]
```diff
@@ -414,3 +427,201 @@ def __call__(self, components: AnimaModularPipeline, state: PipelineState) -> Pi
...
+
+        total_batch = block_state.batch_size * block_state.num_images_per_prompt
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/before_denoise.py]: in modular, we should expand the image_latent in the input step , you can add an image_input step similar to https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/anima/before_denoise.py#L196

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py [hunk 7]
```diff
@@ -122,18 +124,264 @@ def outputs(self):
...
+            "Denoise block for Anima image-to-image generation. "
+            "Uses image_latents already in state from AnimaImg2ImgVaeEncoderStep."
+        )
+
+    @property
+    def outputs(self):
+        return [OutputParam.template("latents")]
+
+
+# auto_docstring
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/modular_blocks_anima.py:203]: why do we need this? can you just add the `set_timesteps` as part of core_denoise?

File: src/diffusers/modular_pipelines/anima/before_denoise.py (hunk 5 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/before_denoise.py]: Hi @yiyixuxu, thank you for the suggestion. 
The get_timesteps() call has been moved into AnimaImg2ImgSetTimestepsStep, which now computes the full schedule and slices it by strength in a single step. AnimaImg2ImgPrepareLatentsStep no longer touches the timestep schedule.

File: src/diffusers/modular_pipelines/anima/before_denoise.py (hunk 6 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/before_denoise.py]: Hi @yiyixuxu, thank you for the suggestion. 
A new AnimaImageInputStep has been added following the same pattern as AnimaTextInputStep - it expands image_latents to batch_size * num_images_per_prompt via repeat_tensor_to_batch_size and derives height/width from the latents when not provided. AnimaImg2ImgPrepareLatentsStep now only handles noise generation and scale_noise().

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py (hunk 7 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/modular_blocks_anima.py:203]: Hi @yiyixuxu, you're right - the wrapper was left over from an earlier structure and is no longer needed. AnimaImg2ImgDenoiseStep has been removed, and AnimaImg2ImgSetTimestepsStep is now part of AnimaImg2ImgCoreDenoiseStep directly, making it a flat sequence (text_conditioning → input → image_input → set_timesteps → prepare_latents → denoise) consistent with ZImageImage2ImageCoreDenoiseStep.

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py [hunk 8]
```diff
@@ -122,18 +125,202 @@ def outputs(self):
...
+class AnimaAutoDenoiseStep(AutoPipelineBlocks):
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/anima/modular_blocks_anima.py]:
```suggestion
class AnimaAutoCoreDenoiseStep(AutoPipelineBlocks):
```

File: tests/modular_pipelines/anima/test_modular_pipeline_anima_img2img.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+)
+
+from ...testing_utils import enable_full_determinism
+from ..test_modular_pipelines_common import ModularPipelineTesterMixin
+
+
+enable_full_determinism()
+
+
+ANIMA_IMG2IMG_WORKFLOWS = {
```
Comment (yiyixuxu) [inline tests/modular_pipelines/anima/test_modular_pipeline_anima_img2img.py]: can you put into same file https://github.com/huggingface/diffusers/blob/main/tests/modular_pipelines/anima/test_modular_pipeline_anima.py

File: src/diffusers/modular_pipelines/anima/modular_blocks_anima.py (hunk 8 above)
Comment (PreethamNoelP) [reply src/diffusers/modular_pipelines/anima/modular_blocks_anima.py]: Hi @yiyixuxu, thank you for the suggestion. 
AnimaAutoDenoiseStep has been renamed to AnimaAutoCoreDenoiseStep, consistent with the naming used across the other modular pipelines.

File: tests/modular_pipelines/anima/test_modular_pipeline_anima_img2img.py (hunk 9 above)
Comment (PreethamNoelP) [reply tests/modular_pipelines/anima/test_modular_pipeline_anima_img2img.py]: Hi @yiyixuxu, done - the img2img test class has been moved into test_modular_pipeline_anima.py and the separate test file has been removed.

**Review by yiyixuxu — APPROVED** [review]
thanks
